akka-http/src/main/scala/akka/http/scaladsl/server/directives/RangeDirectives.scala

```diff
@@ -73,12 +75,31 @@ trait RangeDirectives {
         // Therefore, ranges need to be sorted to prevent that some selected ranges already start to accumulate data
         // but cannot be sent out because another range is blocking the queue.
         val coalescedRanges = coalesceRanges(iRanges).sortBy(_.start)
-        val bodyPartTransformers = coalescedRanges.map(ir ⇒ StreamUtils.sliceBytesTransformer(ir.start, ir.length)).toVector
-        val bodyPartByteStreams = StreamUtils.transformMultiple(entity.dataBytes, bodyPartTransformers)
-        val bodyParts = (coalescedRanges, bodyPartByteStreams).zipped.map { (range, bytes) ⇒
-          Multipart.ByteRanges.BodyPart(range.contentRange(length), HttpEntity(entity.contentType, range.length, bytes))
+        val source = coalescedRanges.size match {
+          case 0 ⇒ Source.empty
+          case 1 ⇒
+            val range = coalescedRanges.head
+            val flow = StreamUtils.sliceBytesTransformer(range.start, range.length)
+            val bytes = entity.dataBytes.via(flow)
+            val part = Multipart.ByteRanges.BodyPart(range.contentRange(length), HttpEntity(entity.contentType, range.length, bytes))
+            Source.single(part)
+          case n ⇒
+            Source fromGraph GraphDSL.create() { implicit b ⇒
+              import GraphDSL.Implicits._
+              val bcast = b.add(Broadcast[ByteString](n))
+              val merge = b.add(Concat[Multipart.ByteRanges.BodyPart](n))
+              for (range ← coalescedRanges) {
+                val flow = StreamUtils.sliceBytesTransformer(range.start, range.length)
+                bcast ~> flow.buffer(16, OverflowStrategy.backpressure).prefixAndTail(0).map {
+                  case (_, bytes) ⇒
+                    Multipart.ByteRanges.BodyPart(range.contentRange(length), HttpEntity(entity.contentType, range.length, bytes))
+                } ~> merge
+              }
+              entity.dataBytes ~> bcast
+              SourceShape(merge.out)
+            }
         }
```

Comment:
@rkuhn Perhaps we can generalize `Split` (combine `splitBefore` and `splitAfter`) to emit multiple sources and get rid of `sliceBytesTransformer`?